dispatcher/nm-dispatcher.c

```diff
@@ -575,41 +575,87 @@ dispatch_one_script (Request *request)
 	return FALSE;
 }
 
-static GSList *
-find_scripts (const char *str_action)
+static int
+_compare_basenames (gconstpointer a, gconstpointer b)
+{
+	const char *basename_a = strrchr (a, '/');
+	const char *basename_b = strrchr (b, '/');
+	int ret;
+
+	g_return_val_if_fail (basename_a, 0);
+	g_return_val_if_fail (basename_b, 0);
+
+	ret = strcmp (++basename_a, ++basename_b);
+	if (ret)
+		return ret;
+
+	return strcmp (a, b);
+}
+
+static void
+_find_scripts (GHashTable *scripts, const char *base, const char *subdir)
 {
-	GDir *dir;
 	const char *filename;
-	GSList *sorted = NULL;
+	gs_free char *dirname = NULL;
 	GError *error = NULL;
-	const char *dirname;
+	GDir *dir;
 
-	if (   strcmp (str_action, NMD_ACTION_PRE_UP) == 0
-	    || strcmp (str_action, NMD_ACTION_VPN_PRE_UP) == 0)
-		dirname = NMD_SCRIPT_DIR_PRE_UP;
-	else if (   strcmp (str_action, NMD_ACTION_PRE_DOWN) == 0
-	         || strcmp (str_action, NMD_ACTION_VPN_PRE_DOWN) == 0)
-		dirname = NMD_SCRIPT_DIR_PRE_DOWN;
-	else
-		dirname = NMD_SCRIPT_DIR_DEFAULT;
+	dirname = g_build_filename (base, "dispatcher.d", subdir, NULL);
 
 	if (!(dir = g_dir_open (dirname, 0, &error))) {
 		g_message ("find-scripts: Failed to open dispatcher directory '%s': %s",
 		           dirname, error->message);
 		g_error_free (error);
-		return NULL;
+		return;
 	}
 
 	while ((filename = g_dir_read_name (dir))) {
-		char *path;
+		g_hash_table_insert (scripts,
+		                     g_strdup (filename),
+		                     g_build_filename (dirname, filename, NULL));
+	}
+
+	g_dir_close (dir);
+}
+
+static GSList *
+find_scripts (const char *str_action)
+{
+	GSList *script_list = NULL;
+	GHashTable *scripts = NULL;
+	GHashTableIter iter;
+	const char *subdir = NULL;
+	char *path;
+	char *filename;
+
+	if (   strcmp (str_action, NMD_ACTION_PRE_UP) == 0
+	    || strcmp (str_action, NMD_ACTION_VPN_PRE_UP) == 0)
+		subdir = "pre-up.d";
+	else if (   strcmp (str_action, NMD_ACTION_PRE_DOWN) == 0
+	         || strcmp (str_action, NMD_ACTION_VPN_PRE_DOWN) == 0)
+		subdir = "pre-down.d";
+
+	scripts = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+	_find_scripts (scripts, NMLIBDIR, subdir);
+	_find_scripts (scripts, NMCONFDIR, subdir);
+
+	g_hash_table_iter_init (&iter, scripts);
+	while (g_hash_table_iter_next (&iter, (gpointer *)&filename, (gpointer *)&path)) {
 		struct stat st;
+		char *link_target;
 		int err;
 		const char *err_msg = NULL;
 
 		if (!check_filename (filename))
```

Comment:
`check_filename()` should be done by `_find_scripts()`. If the filename isn't valid, it can be rejected right away. Then you also don't need the "filename" variable here.

Reply:
Good point.